Gaussian Blur Optimization with Vulkan

iqsxw opened the pull request:
This commit adds a powerful and customizable gblur Vulkan filter,
which provides a maximum 127x127 kernel size of Gaussian Filter.
The size could be adjusted by requirements on quality or performance.

Signed-off-by: Wu Jianhua <jianhua.wu@intel.com>

xuguangxin commented on libavfilter/vf_gblur_vulkan.c:
```diff
@@ -0,0 +1,472 @@
+#define OFFSET(x) offsetof(GBlurVulkanContext, x)
+#define FLAGS (AV_OPT_FLAG_FILTERING_PARAM | AV_OPT_FLAG_VIDEO_PARAM)
+static const AVOption gblur_vulkan_options[] = {
+    { "sigma", "Set sigma", OFFSET(sigma), AV_OPT_TYPE_FLOAT, {.dbl = 0.5}, 0.001, 1024.0, FLAGS },
```
why it's 0.001 not 0?

iqsxw replied:
The reason here is that using zero as the variance does nothing. As for 0.001, it's to avoid the user from entering a too-small value. Otherwise, the subsequent operation's a waste of resources and time.

xuguangxin replied:
If we use 0, it will fail on GPU and success on CPU.

xuguangxin commented on libavfilter/vf_gblur_vulkan.c:
```diff
@@ -0,0 +1,472 @@
+#define FLAGS (AV_OPT_FLAG_FILTERING_PARAM | AV_OPT_FLAG_VIDEO_PARAM)
+static const AVOption gblur_vulkan_options[] = {
+    { "sigma", "Set sigma", OFFSET(sigma), AV_OPT_TYPE_FLOAT, {.dbl = 0.5}, 0.001, 1024.0, FLAGS },
+    { "size", "Set kernel Size(If not set, generated automatically for better quality)", OFFSET(size), AV_OPT_TYPE_INT, {.i64 = 19}, 1, GBLUR_MAX_KERNEL_SIZE, FLAGS },
```
why we need kernel size? cpu version not needed?

iqsxw replied:
The CPU version uses the whole image extent as its kernel size, which is unnecessary to be specified. This is more about the difference between the execution manners of CPU and GPU.

xuguangxin commented on libavfilter/vf_gblur_vulkan.c:
```diff
@@ -0,0 +1,472 @@
+    av_free(kernel_def);
+
+    return 0;
+fail:
```
do we need free kernel_def?

iqsxw replied:
Sure. If not we may have a memory leak here. The memory chunk for kernel_def was allocated in the function, av_asprintf.

iqsxw committed:
libavfilter: add a gblur_vulkan filter

This commit adds a powerful and customizable gblur Vulkan filter,
which provides a maximum 127x127 kernel size of Gaussian Filter.
The size could be adjusted by requirements on quality or performance.

The following command is on how to apply gblur_vulkan filter:

ffmpeg -init_hw_device vulkan -i input.264 -vf hwupload=extra_hw_frames=16,gblur_vulkan,hwdownload,format=yuv420p output.264

Signed-off-by: Wu Jianhua <jianhua.wu@intel.com>

iqsxw commented:
I'll close this pull request, for, the patch has been merged into the upstream,
